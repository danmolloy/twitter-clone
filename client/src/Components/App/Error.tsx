import { GlobeIcon, RefreshIcon } from "@heroicons/react/outline"

export const Error = () => {

  const handleClick = () => {
    alert(Object.keys(localStorage))
    window.location.reload()
  }

  return (
    <div id="error-page" className="w-full flex flex-col items-center mt-12">
      <GlobeIcon className="h-6 w-auto"/>
      <p className="text-gray-600">An error occurred. Please check your connection and try again.</p>
      <button className="tweet-btn py-1 flex flex-row" onClick={() => handleClick()}>
        <RefreshIcon className="h-6 w-auto"/>
        <p className="font-semibold px-1">Retry</p>
      </button>
    </div>
  )
}