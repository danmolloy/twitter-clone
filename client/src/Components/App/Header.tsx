export const Header = (props: any) => {
  return (
    <div className="w-full p-2 border-b border-r">
      <h2 className="text-xl font-semibold">{props.pageTitle}</h2>
      {props.blurb &&
      <p>{props.blurb}</p>}
    </div>
  )
}