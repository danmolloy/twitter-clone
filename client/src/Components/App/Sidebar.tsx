import { HashtagIcon, 
  HomeIcon, 
  UserIcon, 
  BellIcon,
  InboxIcon,
  LogoutIcon,
} from '@heroicons/react/outline'
import { Link } from 'react-router-dom'
import { AUTH_TOKEN } from '../../constants'
import { gql, useQuery } from '@apollo/client'
import { Error } from './Error'
import { Loading } from './Loading'
import { useState } from 'react'
import { UserOptions } from './UserOptions'
import { Message, Notification } from '../../types'

export const DELETE_USER = gql`
  mutation SignUpMutation {
    deleteUser {
      name
    }
  }
`;

export const GET_USER = gql`
  query {
    getNotifications {
      notifications {
          read
        }
        chats {
          content {
            read
            authorHandle
          }
        }
        handle
        profilePic
    }
  }
`;


export const Sidebar = () => {
  const [showUserOptions, setShowUserOptions] = useState(false)

  const { loading, error, data } = useQuery(GET_USER)

  const onSignOut = () => {
    localStorage.removeItem(AUTH_TOKEN)
    window.location.reload()
  }

  if (error) {
    console.log(error)
    return <Error />
  }

  if (loading) {
    return <Loading />
  }

  return (
    <div id="side-bar" 
    className="side-bar">
      <Link to="/home" className="md:ml-2" id="twitter-home-link" title="Home">
        <img src="/favico.ico" alt="Blue Twitter bird" className="hidden sm:flex w-12 h-auto mx-6 my-4 p-2 rounded hover:bg-blue-50"/>
      </Link>
      <Link to="/home" className="lg-side-icon" id="home-link" title="Home">
        <HomeIcon className="side-icon"/>
        <label htmlFor="home-link" className="sidebar-text">Home</label>
      </Link>
      <Link to="/explore" title="Explore" className="lg-side-icon" id="explore-link">
        <HashtagIcon className="side-icon"/>
        <label htmlFor="explore-link" className="sidebar-text">Explore</label>
      </Link>
      <Link to="/notifications" title="Notifications" className="lg-side-icon" id="notifications-link">
      {data && data.getNotifications.notifications.map(
        (i: Notification) => i.read).includes(false) 
      && <div id="notifications-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <BellIcon className="side-icon"/>
        <label htmlFor="notifications-link" className="sidebar-text">Notifications</label>
      </Link>
      <Link to="/messages" title="Messages" className="lg-side-icon">
      {data && data.getNotifications.chats[0] !== undefined
      && data.getNotifications.chats[0].content.filter(
        (i:Message) => i.read === false && i.authorHandle !== data.getNotifications.handle).length > 0
        && <div id="unread-msg-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <InboxIcon className="side-icon"/>
        <label htmlFor="messages-link" className="sidebar-text">Messages</label>
      </Link>
      <button className="sm:hidden lg-side-icon" title="User options" onClick={() => setShowUserOptions(!showUserOptions)} >
        {showUserOptions 
        && <UserOptions 
        currentUser={data && data.getNotifications} 
        onSignOut={() => onSignOut()} 
        close={() => setShowUserOptions(false)}/>}
        <img alt={data.getNotifications?.name ? data.getNotifications?.name : "Current user"} src={data && data.getNotifications?.profilePic} className="side-icon rounded-full"/>
      </button>
      <Link id="profile-link" to={data ? `/${data.getNotifications.handle.slice(1)}`: '/'} title="Profile" className="hidden sm:flex lg-side-icon">
        <UserIcon className="side-icon hidden sm:flex"/>
        <label htmlFor="profile-link" className="sidebar-text">Profile</label>
      </Link>
      <Link title="Sign out" id="sign-out-btn" className="hidden sm:flex lg-side-icon" to="/home" onClick={() => onSignOut()}>
        <LogoutIcon className="side-icon hidden sm:flex" id="more-icon"/>
        <label htmlFor="sign-out" className="sidebar-text">Sign Out</label>
      </Link>
    </div>
  )
}