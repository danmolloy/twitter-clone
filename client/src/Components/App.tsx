import React, { useEffect, useState } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  Redirect
} from "react-router-dom"
import { Home } from './Home'
import { Explore } from './Explore'
import { Bookmarks } from './Bookmarks'
import { Messages } from './Messages'
import { Notifications } from './Notifications'
import { Profile } from './Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "./ComposeTweet";
import { RightBar }  from './RightBar'
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../constants'
import { Chat } from "./Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    follows {
      handle
    }
    followers {
      handle
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar currentUser={data && data.currentUser}/>}
        <div className="sm:ml-24 md:ml-60 mb-0 border-r w-full max-w-2xl sm:mr-2">
        <Switch>
          <Route path="/compose/tweet">
            <ComposeTweet currentUser={data && data.currentUser} updatePage={() => alert('Posted!')}/>
          </Route>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/bookmarks">
            <Bookmarks currentUser={data && data.currentUser} />
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
          <RightBar />
        </div>
    </Router>
  );
}

export default App;
