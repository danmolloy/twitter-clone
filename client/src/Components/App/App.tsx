import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from "react-router-dom"
import { Home } from '../Home/Home'
import { Explore } from '../Explore/Explore'
import { Messages } from '../Messages/Messages'
import { Notifications } from '../Notifications/Notifications'
import { Profile } from '../Profile/Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "../Home/ComposeTweet";
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../../constants'
import { Chat } from "../Messages/Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    chats {
      content {
        read
      }
    }
    follows {
      handle
    }
    followers {
      handle
    }
    notifications {
      id
      text
      sentFromUser
      read
      tweetId
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    console.log(error)
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar />}
        <div className="sm:ml-24 md:ml-60 mb-0 w-full max-w-2xl sm:mr-2 md:mr-0">
        <Switch>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
      </div>
    </Router>
  );
}

export default App;
