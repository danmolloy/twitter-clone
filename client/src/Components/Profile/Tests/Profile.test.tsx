import { act } from "react-dom/test-utils";
import { unmountComponentAtNode, render } from "react-dom";
import { MemoryRouter, Route } from "react-router";
import { MockedProvider } from "@apollo/client/testing";
import { AUTH_TOKEN } from "../../../constants";
import pretty from "pretty";
import { fireEvent } from "@testing-library/dom";
import { ProfileMock } from './ProfileMock'
import App from "../../App/App";
import { getByText } from "@testing-library/dom";


let container: any = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
  global.localStorage.setItem(AUTH_TOKEN, "auth-token");
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  global.localStorage.removeItem(AUTH_TOKEN);
});

it("renders without error", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(pretty(container.innerHTML)).toMatchSnapshot();
})

it("arrow button links to home", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-home-link"))
  expect(pretty(container.innerHTML)).toMatch(/id="home/g)
})

it("shows following", async() => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).not.toMatch(/FollowingGeorge CostanzaDan MolloyFind more users in Explore./g)
  fireEvent.click(container.querySelector("#following-btn"))
  expect(container.textContent).toMatch(/FollowingGeorge CostanzaDan MolloyFind more users in Explore./g)
  // Test it closes window
})

it("shows followers", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).not.toMatch(/FollowersDan Molloy/g)
  fireEvent.click(container.querySelector("#followers-btn"))
  expect(container.textContent).toMatch(/FollowersDan Molloy/g)
  // Test it closes window
})

it("filters retweets and tweets", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).toMatch(/A profound love between two people/g)
  expect(container.textContent).not.toMatch(/You haven't retweeted any tweets./g)

  fireEvent.click(container.querySelector("#profile-retweets-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })  
  expect(container.textContent).not.toMatch(/A profound love between two people/g)
  expect(container.textContent).toMatch(/You haven't retweeted any tweets./g)
})

it("likes tweet in profile page", async () => {
  let testHistory, testLocation
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
          <Route 
          path="*"
          render={({ history, location }) => {
            testHistory = history;
            testLocation = location;
            return null;
          }}
          />
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
 await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.querySelector("like-count")).toBe(null)
  fireEvent.click(container.querySelector("#like-button"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.querySelector("#like-count").textContent).toMatch(/^1$/)
})

it("retweets post in profile page", async () => {
  jest.spyOn(window, 'alert').mockImplementation(() => {});
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
 await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.querySelector("#retweet-count").textContent).not.toMatch(/^1$/)
  fireEvent.click(container.querySelector("#retweet-button"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(window.alert).toHaveBeenCalled()
  expect(container.querySelector("#retweet-count").textContent).toMatch(/^1$/)
   fireEvent.click(container.querySelector("#profile-retweets-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  }) 
  expect(container.textContent).toMatch(/A profound love between two people/g)
  expect(container.textContent).not.toMatch(/You haven't retweeted any tweets./g)
})

it("posts comment without error", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
 await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.querySelector("#comments-count").textContent).not.toMatch(/^1$/g)
  expect(pretty(container.querySelector("#single-tweet").innerHTML)).not.toMatch(/Tweet your reply/g)

  fireEvent.click(container.querySelector("#tweet-comments-btn"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(pretty(container.querySelector("#single-tweet").innerHTML)).toMatch(/Tweet your reply/g)
  fireEvent.change(container.querySelector("#comment-input"), {target: {value: "A comment from Jest!"}})
  fireEvent.click(container.querySelector("#post-comment-btn"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.querySelector("#comments-count").textContent).toMatch(/^1$/g)
})

it("deletes post", async () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  fireEvent.click(container.querySelector("#profile-link"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  container.querySelector(".post-options").focus()
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).toMatch(/Delete Tweet/g)
  expect(container.textContent).toMatch(/A profound love /g)
  fireEvent.click(container.querySelector("#delete-button"))
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).not.toMatch(/A profound love /g)
  expect(container.textContent).toMatch(/You haven't posted any tweets./g)
})


/*it("follows & unfollows user from tweet options", () => {
})

it("follows and unfollows user from follow btn", () => {})

it("Edits profile without error", () => {
  act(() => {
    render(
      <MockedProvider mocks={ProfileMock} addTypename={false}>
        <MemoryRouter>
          <App /> 
        </MemoryRouter>
      </MockedProvider>, container
    )
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  act(() => {
    fireEvent.click(container.querySelector("#profile-link"))
  })
  await act(async() => {
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  act(() => {
    fireEvent.click(container.querySelector("#edit-profile-btn"))
  })
  expect(container.textContent).toMatch(/Profile Pic HTMLFull NameBlurbSubmit/g)
  act(() => {
    fireEvent.change(container.querySelector("#profile-pic-input"), {target: {value: "jest.jpg"}})
  })
  act(() => {
    fireEvent.change(container.querySelector("#name-input"), {target: {value: "Jesty Jest"}})
  })
  act(() => {
    fireEvent.change(container.querySelector("#blurb-input"), {target: {value: "Jest changed the blurb!"}})
  })
  await act(async () => {
    fireEvent.click(container.querySelector("#submit-edit-btn"))
    await new Promise(resolve => setTimeout(resolve, 100))
  })
  expect(container.textContent).toMatch(/Sign OutJesty Jest/g)
  expect(container.textContent).toMatch(/Jest changed the blurb!Joined Nov 24 /g)
  expect(container.querySelector("#profile-pic").src).toMatch(/jest.jpg/g)
})

  */