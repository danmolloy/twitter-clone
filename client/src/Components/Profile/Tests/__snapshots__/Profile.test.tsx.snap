// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without error 1`] = `
"<div>
  <div id=\\"main-content\\" class=\\"flex flex-row w-screen h-screen justify-start\\">
    <div id=\\"side-bar\\" class=\\"side-bar\\"><a class=\\"md:ml-2\\" id=\\"twitter-home-link\\" title=\\"Home\\" href=\\"/home\\"><img src=\\"/favico.ico\\" alt=\\"Blue Twitter bird\\" class=\\"hidden sm:flex w-12 h-auto mx-6 my-4 p-2 rounded hover:bg-blue-50\\"></a><a class=\\"lg-side-icon\\" id=\\"home-link\\" title=\\"Home\\" href=\\"/home\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6\\"></path>
        </svg><label for=\\"home-link\\" class=\\"sidebar-text\\">Home</label></a><a title=\\"Explore\\" class=\\"lg-side-icon\\" id=\\"explore-link\\" href=\\"/explore\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M7 20l4-16m2 16l4-16M6 9h14M4 15h14\\"></path>
        </svg><label for=\\"explore-link\\" class=\\"sidebar-text\\">Explore</label></a><a title=\\"Notifications\\" class=\\"lg-side-icon\\" id=\\"notifications-link\\" href=\\"/notifications\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9\\"></path>
        </svg><label for=\\"notifications-link\\" class=\\"sidebar-text\\">Notifications</label></a><a title=\\"Messages\\" class=\\"lg-side-icon\\" href=\\"/messages\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M20 13V6a2 2 0 00-2-2H6a2 2 0 00-2 2v7m16 0v5a2 2 0 01-2 2H6a2 2 0 01-2-2v-5m16 0h-2.586a1 1 0 00-.707.293l-2.414 2.414a1 1 0 01-.707.293h-3.172a1 1 0 01-.707-.293l-2.414-2.414A1 1 0 006.586 13H4\\"></path>
        </svg><label for=\\"messages-link\\" class=\\"sidebar-text\\">Messages</label></a><button class=\\"sm:hidden lg-side-icon\\" title=\\"User options\\"><img alt=\\"Current user\\" src=\\"profilePic.jpg\\" class=\\"side-icon rounded-full\\"></button><a id=\\"profile-link\\" title=\\"Profile\\" class=\\"hidden sm:flex lg-side-icon\\" href=\\"/leGuin\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon hidden sm:flex\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M16 7a4 4 0 11-8 0 4 4 0 018 0zM12 14a7 7 0 00-7 7h14a7 7 0 00-7-7z\\"></path>
        </svg><label for=\\"profile-link\\" class=\\"sidebar-text\\">Profile</label></a><a title=\\"Sign out\\" id=\\"sign-out-btn\\" class=\\"hidden sm:flex lg-side-icon\\" href=\\"/home\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon hidden sm:flex\\" id=\\"more-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1\\"></path>
        </svg><label for=\\"sign-out\\" class=\\"sidebar-text\\">Sign Out</label></a></div>
    <div id=\\"app-page\\" class=\\"sm:ml-24 md:ml-60 mb-0 w-full max-w-2xl sm:mr-2 md:mr-0\\">
      <div id=\\"profile-component\\" class=\\"border-r w-full mr-2\\">
        <div id=\\"profile-header\\" class=\\"border-b flex flex-row\\"><a id=\\"profile-home-link\\" href=\\"/home\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"w-10 p-2 h-auto ml-4 my-2\\">
              <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M10 19l-7-7m0 0l7-7m-7 7h18\\"></path>
            </svg></a>
          <div class=\\" ml-2 mt-1\\">
            <h3 class=\\"text-xl font-bold\\">Ursula LeGuin</h3>
            <p class=\\"text-sm text-gray-600 -mt-1\\">1 tweets</p>
          </div>
        </div>
        <div id=\\"profile-details\\" class=\\"w-full h-3/5 \\"><img id=\\"profile-banner\\" alt=\\"Default profile background\\" class=\\"w-full h-48\\" src=\\"/bgPic.jpg\\">
          <div class=\\"flex flex-row justify-between\\"><img id=\\"profile-pic\\" src=\\"profilePic.jpg\\" alt=\\"Profile picture of Ursula LeGuin\\" class=\\"rounded-full w-36 h-auto ml-4 -mt-12 -p border-4 border-white\\">
            <div class=\\"flex flex-col\\"><button id=\\"edit-profile-btn\\" class=\\"border tweet-btn\\">Edit profile</button></div>
          </div>
          <div class=\\"flex flex-col w-4/5 sm:w-2/5 ml-12\\">
            <h3 class=\\"font-bold text-xl\\">Ursula LeGuin</h3>
            <p class=\\"text-gray-600 text-sm\\">@leGuin</p>
            <p class=\\"my-2\\">Click Edit Profile!</p>
            <div class=\\"flex flex-row text-gray-600\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"w-6 h-auto -ml-1\\">
                <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M8 7V3m8 4V3m-9 8h10M5 21h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v12a2 2 0 002 2z\\"></path>
              </svg>
              <p class=\\"ml-1\\">Joined Nov 24 2021</p>
            </div>
            <div id=\\"profile-followers\\" class=\\"flex flex-row text-gray-600 my-2\\">
              <div><button id=\\"following-btn\\"><span class=\\"font-bold text-black\\">2</span> following</button></div>
              <div><button id=\\"followers-btn\\"><span class=\\"font-bold text-black ml-2\\">1</span> followers</button></div>
            </div>
          </div>
        </div>
        <div id=\\"profile-tweets\\" class=\\"w-full flex flex-col items-center min-h-screen\\">
          <div class=\\"w-full h-12 mt-4 flex flex-row\\"><a id=\\"profile-tweets-link\\" class=\\"selected-tweet-filter\\" href=\\"/leGuin\\">Tweets<span class=\\"tab-line\\"></span></a><a id=\\"profile-retweets-link\\" class=\\"deselected-tweet-filter\\" href=\\"/leGuin/retweets\\">Retweets</a></div>
          <div id=\\"profile-tweets\\" class=\\"h-full w-full flex flex-col mt-0\\">
            <div id=\\"single-tweet\\" class=\\"border-b hover:bg-gray-50 flex flex-col items-center\\">
              <div class=\\"flex flex-row mt-4 w-full\\"><a class=\\"user-profile-link\\" href=\\"/leGuin\\"><img src=\\"profilePic.jpg\\" class=\\"w-14 h-auto mx-3 rounded-full\\"></a>
                <div class=\\"ml-3 flex flex-col w-full\\">
                  <div class=\\"flex flex-row w-full justify-between\\">
                    <div class=\\"flex flex-col sm:flex-row \\"><a class=\\"user-profile-link flex flex-row flex-wrap\\" href=\\"/leGuin\\">
                        <h3 class=\\"font-bold hover:underline\\">Ursula LeGuin</h3>
                        <h4 class=\\"text-gray-500 ml-1\\">@leGuin</h4>
                      </a><span class=\\"hidden sm:flex text-gray-500 ml-1\\">•</span>
                      <h4 class=\\"text-gray-500 ml-1 hover:underline\\">Tue Sep 28 2021</h4>
                    </div>
                    <div id=\\"tweet-options\\"><button id=\\"post-options-btn\\" class=\\"post-options text-gray-500 hover:bg-blue-50 hover:text-blue-500 rounded-full p-1 mr-2\\">•••</button></div>
                  </div>
                  <div>
                    <p class=\\" mr-2\\">A profound love between two people involves, after all, the power and chance of doing profound hurt.</p>
                  </div>
                </div>
              </div>
              <div class=\\"flex flex-row justify-between px-12 my-2 text-gray-500 w-full\\"><button id=\\"tweet-comments-btn\\" class=\\"show-comments flex flex-row items-center hover:text-blue-500\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\" hover:bg-blue-50 tweet-options\\">
                    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M8 12h.01M12 12h.01M16 12h.01M21 12c0 4.418-4.03 8-9 8a9.863 9.863 0 01-4.255-.949L3 20l1.395-3.72C3.512 15.042 3 13.574 3 12c0-4.418 4.03-8 9-8s9 3.582 9 8z\\"></path>
                  </svg>
                  <p id=\\"comments-count\\">0</p>
                </button><button id=\\"retweet-button\\" class=\\"flex flex-row items-center hover:text-green-500\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"hover:bg-green-50 tweet-options\\">
                    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15\\"></path>
                  </svg>
                  <p id=\\"retweet-count\\"></p>
                </button><button id=\\"like-button\\" class=\\"flex flex-row  items-center hover:text-red-500\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"hover:bg-red-50 tweet-options\\">
                    <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z\\"></path>
                  </svg>
                  <p id=\\"like-count\\"></p>
                </button></div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
